client/wsclient.go

```diff
@@ -245,19 +250,33 @@ func (c *wsClient) runOneCycle(ctx context.Context) {
 		c.common.PackagesStateProvider,
 		c.common.Capabilities,
 	)
-	r.ReceiverLoop(ctx)
-
-	// Stop the background processors.
-	procCancel()
-
-	// If we exited receiverLoop it means there is a connection error, we cannot
-	// read messages anymore. We need to start over.
+	r.Start(ctx)
+
+	select {
+	case <-c.sender.IsStopped():
+		// sender will send close message to initiate the close handshake
+		if err := c.sender.Err(); err != nil && err != websocket.ErrCloseSent {
+			c.common.Logger.Debugf("failed to send close message, close without the handshake.")
+			break
+		}
 
-	// Close the connection to unblock the WSSender as well.
-	_ = c.conn.Close()
+		c.common.Logger.Debugf("waiting for close message from server.")
+		select {
+		case <-r.IsStopped():
+			c.common.Logger.Debugf("shutdown handshake complete.")
+		case <-time.After(c.connShutdownTimeout):
+			c.common.Logger.Debugf("timeout waiting for close message.")
+			// not receive close message from the server, close the connection to force the receive loop to stop
+			_ = c.conn.Close()
+			<-r.IsStopped()
+		}
+	case <-r.IsStopped():
+		// If we exited receiverLoop it means there is a connection error, we cannot
+		// read messages anymore. We need to start over.
 
-	// Wait for WSSender to stop.
-	c.sender.WaitToStop()
+		stopSender()
+		<-c.sender.IsStopped()
```

Comment:
Running `stopSender` will attempt to send a message to the server. Do you know whether this will time out in a reasonable timeframe, or should we race this with `c.connShutdownTimeout` to ensure it doesn't hang?

Reply:
`c.connShutdownTimeout` is controlling how long the wsclient will wait for the ACK from server, and the timeout for sendding the close message to the server is [`defaultSendCloseMessageTimeout`](https://github.com/open-telemetry/opamp-go/pull/213/files#diff-1a7764ac131862565b9f9783634aa0c05d383cc44c34268308865908e1c3d02dR16)(5s). So it will not hang here.

`c.connShutdownTimeout` is currently only modified during testing and both of them are not configurable for now. We can improve the Client API to make them configurable later.